File: src/main/java/puzzles/year23/day12/Part2.java
package puzzles.year23.day12;

import util.Puzzle;

import java.util.*;
import java.util.regex.Pattern;

public class Part2 implements Puzzle<Long> {

    @Override
    public Long getAnswer(String inputString) {
        String[] lines = inputString.split(System.lineSeparator());
        long total = 0L;

        int start = 12;
        int count = 0;

        for (String line : lines) {
            if (count <= start) {
                count++;
                continue;
            }
            String[] lineComponents = line.split(" ");
            String pattern = lineComponents[0] + "?" + lineComponents[0] + "?" + lineComponents[0] + "?" + lineComponents[0] + "?" + lineComponents[0];
//            String pattern = lineComponents[0];
//            String repeatString = lineComponents[1];
            String repeatString = lineComponents[1] + "," + lineComponents[1] + "," + lineComponents[1] + "," + lineComponents[1] + "," + lineComponents[1];
            int[] repeat = Arrays.stream(repeatString.split(",")).mapToInt(Integer::parseInt).toArray();

            Queue<String> queue = new LinkedList<>();
            queue.add(pattern);

            // Iterate through the strings
            // Have a queue to add things to
            // For every ., reset counter
            // For every # increase counter
            // For every ?, if you need a # place it
            // If counter is 0, and there is a ?, need to branch
            int runningTotal = 0;
            while (!queue.isEmpty()) {
                char[] input = queue.poll().toCharArray();
                int counter = 0;
                int patternIndex = 0;
                boolean broken = false;
                for (int i = 0; i < input.length; i++) {
                    char c = input[i];

                    if (c == '.') {
                        if (counter != 0) {
                            if (counter != repeat[patternIndex]) {
                                broken = true;
                                break;
                            }
                            patternIndex++;
                        }
                        counter = 0;
                    } else if (c == '#') {
                        if (patternIndex >= repeat.length) {
                            broken = true;
                            break;
                        }
                        counter++;
                        if (counter > repeat[patternIndex]) {
                            broken = true;
                            break;
                        }
                    } else if (c == '?') {
                        if (patternIndex >= repeat.length) {
                            input[i] = '.';
                            counter = 0;
                        } else if (counter == 0) {
                            input[i] = '.';
                            queue.add(new String(input));
                            input[i] = '#';
                            counter++;
                            if (counter > repeat[patternIndex]) {
                                broken = true;
                                break;
                            }
                        } else if (counter < repeat[patternIndex]) {
                            input[i] = '#';
                            counter++;
                            if (counter > repeat[patternIndex]) {
                                broken = true;
                                break;
                            }
                        } else if (counter == repeat[patternIndex]) {
                            input[i] = '.';
                            patternIndex++;
                            counter = 0;
                        }
                    }
                }

                if (input[input.length - 1] == '#' && repeat[repeat.length - 1] == counter) {
                    patternIndex++;
                }

                if (!broken && patternIndex == repeat.length) {

                    // Got through the whole string
//                    System.out.println(Arrays.toString(input));
//                    System.out.println("Pattern Index: " + patternIndex + ", repeat: " + repeat.length);
                    runningTotal++;
                }
            }
            total += runningTotal;
            System.out.println(count++ + ": " + runningTotal);
        }

        return total;
    }
}

File: src/main/java/puzzles/year24/day19/Part2.kt
package puzzles.year24.day19

import util.Puzzle

class Part2 : Puzzle<Long?> {
    val cache = mutableMapOf<String, Long>()

    override fun getAnswer(inputString: String): Long {
        val towels = inputString.split(System.lineSeparator() + System.lineSeparator())[0].split(", ")
        val patterns = inputString.split(System.lineSeparator() + System.lineSeparator())[1].lines()

        var result = 0L
        var count = 0
        for (pattern in patterns) {
            cache.clear()
           val newCount = getCount(towels, pattern)
//            println("$pattern -> $newCount")
            result += newCount
            if (newCount > 0) {
                count++
            }
        }

        return result
    }

    private fun getCount(towels: List<String>, pattern: String): Long {
        if (pattern.isBlank()) return 1L

        if (cache.containsKey(pattern)) {
            return cache[pattern]!!
        }

        var totalCount = 0L
        for (towel in towels) {
            if (pattern.startsWith(towel)) {
                val newCount = getCount(towels, pattern.substring(towel.length))
                totalCount += newCount
            }
        }

        cache[pattern] = totalCount
        return totalCount
    }
}
